### ISSUE TITLE
Failure to correctly apply fixes for clippy::map_entry

### ISSUE THREAD

**Issue by hkBst** [rust-lang/rust-clippy#15307]
### Description

Failure to correctly apply fixes for this lint:
```
warning: usage of `contains_key` followed by `insert` on a `HashMap`
   --> src/tools/unicode-table-generator/src/raw_emitter.rs:343:13
    |
343 | /             if !unique_mapping.contains_key(&w) {
344 | |                 assert!(
345 | |                     unique_mapping
346 | |                         .insert(w, UniqueMapping::Canonical(canonical_words.len()))
...   |
349 | |                 canonical_words.push(w);
350 | |             }
    | |_____________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_entry
    = note: `#[warn(clippy::map_entry)]` on by default
help: try
    |
343 ~             if let std::collections::hash_map::Entry::Vacant(e) = unique_mapping.entry(w) {
344 +                 assert!(
345 +                     { e.insert(UniqueMapping::Canonical(canonical_words.len())); None }
346 +                         .is_none()
347 +                 );
348 +                 canonical_words.push(w);
349 +             }
```

Original code:
```
        for &w in unique_words {
            if !unique_mapping.contains_key(&w) {
                assert!(
                    unique_mapping
                        .insert(w, UniqueMapping::Canonical(canonical_words.len()))
                        .is_none()
                );
                canonical_words.push(w);
            }
        }
```

Invocation and output after failure:
```
~/rust/src/tools/unicode-table-generator$ cargo clippy --fix --bin "unicode-table-generator"
    Checking unicode-table-generator v0.1.0 (/home/gh-hkBst/rust/src/tools/unicode-table-generator)
warning: failed to automatically apply fixes suggested by rustc to crate `unicode_table_generator`

after fixes were automatically applied the compiler reported errors within these files:

  * src/tools/unicode-table-generator/src/raw_emitter.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0282]: type annotations needed
   --> src/tools/unicode-table-generator/src/raw_emitter.rs:345:82
    |
345 |                     { e.insert(UniqueMapping::Canonical(canonical_words.len())); None }
    |                                                                                  ^^^^ cannot infer type of the type parameter `T` declared on the enum `Option`
    |
help: consider specifying the generic argument
    |
345 |                     { e.insert(UniqueMapping::Canonical(canonical_words.len())); None::<T> }
    |                                                                                      +++++

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0282`.
Original diagnostics will follow.
```

### Version

```text
rustc 1.90.0-nightly (e46629662 2025-07-17)
binary: rustc
commit-hash: e4662966273ed58b51f9ff8d682accc202aa1210
commit-date: 2025-07-17
host: x86_64-unknown-linux-gnu
release: 1.90.0-nightly
LLVM version: 20.1.8
```

**Comment by ada4a** [comment]
Just to save time of whoever is going to look at this next:

The faulty suggestion is created at the following codepath

https://github.com/rust-lang/rust-clippy/blob/72243f469d04925e63cf3a8e9d459c3254780b35/clippy_lints/src/entry.rs#L664-L668

which appears to not have been covered by any of the existing tests.

I minimized the reproducer even further to just this:
```rs
fn issue15307<K, V>(m: HashMap<K, V>, k: K, v: V) {
    if !m.contains_key(&k) {
        //~^ map_entry
        assert!(m.insert(k, v).is_none());
    }
}
```

which gets the following suggestion:
```rs
LL ~     if let std::collections::hash_map::Entry::Vacant(e) = m.entry(k) {
LL +         assert!({ e.insert(v); None }.is_none());
LL +     }
```

I'm not really sure how to fix the suggestion though. Not linting at all would probably be better than linting incorrectly, but that's not ideal of course.
